### ISSUE TITLE
[JENKINS-12290] Deadlock/Lockup when using "trigger/call builds on other projects" and "Block until the triggered projects finish their builds" is used with only 1 executor

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/parameterized-trigger-plugin#508]
<p>On a jenkins node that has 1 executor that is trying to execute an upstream job via the "trigger/call builds on other projects" build step, the job hangs indefinately or until the user-supplied timeout is reached.</p>

<p>In the console for the main job, it shows:</p>

<p>Waiting for the completion of <name-of-upstream-project></p>

<p>While the Jenkins dashboard shows the upstream project is trying to build, with the text:<br/>
(pending - Waiting for next available executor )</p>

<p>I would expect that the plugin would somehow be able to indicate to the jenkins node that only 1 executor is required.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/garen.png" title="garen's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=garen">garen</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290" target="_blank">Deadlock/Lockup when using "trigger/call builds on other projects" and "Block until the triggered projects finish their builds" is used with only 1 executor</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=huybrechts">huybrechts</a>
<li><b>status</b>: Open
<li><b>priority</b>: Critical
<li><b>component(s)</b>: parameterized-trigger-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 38
<li><b>watchers</b>: 41
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>On a jenkins node that has 1 executor that is trying to execute an upstream job via the "trigger/call builds on other projects" build step, the job hangs indefinately or until the user-supplied timeout is reached.</p>

<p>In the console for the main job, it shows:</p>

<p>Waiting for the completion of &lt;name-of-upstream-project&gt;</p>

<p>While the Jenkins dashboard shows the upstream project is trying to build, with the text:
(pending - Waiting for next available executor )</p>

<p>I would expect that the plugin would somehow be able to indicate to the jenkins node that only 1 executor is required.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/awattar.jpg" title="awattar's avatar" /> awattar:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=338042&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338042">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm also looking forward to see this resolved.</p>
</pre>
</details>

</ul>
</details>

<p>I'm also looking forward to see this resolved.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tigerc:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=351873&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351873">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Also following and in agreement with Vasily that it is expected that if the childjob is asking for a node that is run by the parent node, that it should be allowed access to the node. Without this expectation working, our complex pipeline is unable to properly archive artifacts in the ideal clean manner it was designed for. The alternative is to seek for a solution to archive from nodes</p>
</pre>
</details>

</ul>
</details>

<p>Also following and in agreement with Vasily that it is expected that if the childjob is asking for a node that is run by the parent node, that it should be allowed access to the node. Without this expectation working, our complex pipeline is unable to properly archive artifacts in the ideal clean manner it was designed for. The alternative is to seek for a solution to archive from nodes</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ptahchiev:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=354099&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354099">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've just hit this issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/> My scenario is the following. I have several projects:</p>

<p> - bom</p>

<p> - platform</p>

<p> - archetype</p>

<p> - console</p>

<p> - release-all</p>

<p> </p>

<p>The release-all is a pipeline build which calls release on each of them in the following order bom-&gt;platform-&gt;archetype-&gt;console However because I have just one executor running release-all blocks the executor and bom is never started to release because it is waiting for the next available executor.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>I've just hit this issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/> My scenario is the following. I have several projects:</p>

<p> - bom</p>

<p> - platform</p>

<p> - archetype</p>

<p> - console</p>

<p> - release-all</p>

<p> </p>

<p>The release-all is a pipeline build which calls release on each of them in the following order bom->platform->archetype->console However because I have just one executor running release-all blocks the executor and bom is never started to release because it is waiting for the next available executor.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tamashegedus:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=379439&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379439">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have a single jenkins server with two executors. I cannot run two pipelines in parallel because it immediately gets deadlocked as it cannot start the child jobs. I wonder why doesn't this issue have critical priority. It renders pipeline jobs useless in most cases. <a href="https://issues.jenkins.io/browse/JENKINS-26959" title="Waiting Job blocks executor causing deadlock" class="issue-link" data-issue-key="JENKINS-26959"><del>JENKINS-26959</del></a> was set critical and was closed as a duplicate of this issue.</p>
</pre>
</details>

</ul>
</details>

<p>I have a single jenkins server with two executors. I cannot run two pipelines in parallel because it immediately gets deadlocked as it cannot start the child jobs. I wonder why doesn't this issue have critical priority. It renders pipeline jobs useless in most cases. <a href="https://issue-redirect.jenkins.io/issue/26959" title="Waiting Job blocks executor causing deadlock" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26959">JENKINS-26959</a>"><del>JENKINS-26959</del></a> was set critical and was closed as a duplicate of this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>biodranik:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=383263&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383263">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is a critical issue that makes parallel/matrix builds unusable. In my pipeline I need to build for Mac and Linux. Obviously, it can be done in parallel. There are two build nodes available (Linux and Mac). The bug reproduces when one build node waits until another build node finishes. And another one waits until the first one finishes. This is a deadlock.</p>
</pre>
</details>

</ul>
</details>

<p>This is a critical issue that makes parallel/matrix builds unusable. In my pipeline I need to build for Mac and Linux. Obviously, it can be done in parallel. There are two build nodes available (Linux and Mac). The bug reproduces when one build node waits until another build node finishes. And another one waits until the first one finishes. This is a deadlock.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nhatkhai.png" title="nhatkhai's avatar" /> nhatkhai:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=409818&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409818">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What if we have the queue scheduler reserve x number of executors per agent - so that parent jobs don't block all of executors out from the child (a temporary safe guard)?</p>
</pre>
</details>

</ul>
</details>

<p>What if we have the queue scheduler reserve x number of executors per agent - so that parent jobs don't block all of executors out from the child (a temporary safe guard)?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nhatkhai.png" title="nhatkhai's avatar" /> nhatkhai:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=409819&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409819">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>And/or while parent job wait for the child jobs, it should release the executor. Then when come back to parent jobs, it just had to acquire the new executor again on the same agent machine.</p>

<p>(Or hang over the executor to child jobs, and take back from it when it done)</p>
</pre>
</details>

</ul>
</details>

<p>And/or while parent job wait for the child jobs, it should release the executor. Then when come back to parent jobs, it just had to acquire the new executor again on the same agent machine.</p>

<p>(Or hang over the executor to child jobs, and take back from it when it done)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nhatkhai.png" title="nhatkhai's avatar" /> nhatkhai:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=409821&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409821">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think similar release, and acquire could be apply to all the do no thing operation like: sleep x sec, wait for other job finished etc...</p>
</pre>
</details>

</ul>
</details>

<p>I think similar release, and acquire could be apply to all the do no thing operation like: sleep x sec, wait for other job finished etc...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jcabo.svg" title="jcabo's avatar" /> jcabo:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12290?focusedId=433793&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-433793">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I see some issues with this ticket:</p>
<ul>
	<li>I'm not very familiar about the project/components terms used in Jenkins, however I believe the issue is not about the "parameterized-trigger-plugin" component, but rather about the "build step" used to launch other jobs from within another job.</li>
	<li>The issue is not related to having only 1 executor, but rather to having no available executors when launching too many master/orchestrators jobs that run sub-jobs, causing a deadlock.</li>
</ul>


<p>I propose updating the Jira ticket to reflect the issue as: "Deadlock/Lockup when using "launching builds on other jobs" and "Block until the triggered jobs finish their builds" when no available executors"</p>

<p>From what I understand, this issue occurs in the following scenario:</p>
<ul>
	<li>Node/slave with only 2 (or 1) executors available</li>
	<li>1 upstream job finishes successfully</li>
	<li>2 downstream (orchestrator) jobs are automatically triggered immediatelly by the upstream job when it finishes successfully</li>
	<li>Each downstream job launches other sub-jobs (dependent jobs) that each require their own executor</li>
</ul>


<p>Since the last bullet point requires extra executors, and there are none available, the master/orchestrator jobs will enter a deadlock.</p>

<p>There are several ways to resolve this issue:</p>
<ol>
	<li>(my preference): The launched dependent jobs do not occupy extra executors if run on the same node/slave.</li>
	<li>Distribute the execution of the downstream jobs so that they do not collide and there are available executors for their sub-jobs.</li>
</ol>


<p>A high-level implementation of the above could be:</p>
<ol>
	<li>For the first option, this could be automated by the Jenkins orchestrator. Alternatively, a new flag in the build step could also help.</li>
	<li>For the second way, this could be achieved by adding a flag to the trigger-upstream configuration in the downstream job with a parameter similar to the cron H flag. In other words, instead of immediately executing the downstream jobs, they could be distributed over a set period of time (e.g. within an hour).</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>I see some issues with this ticket:</p>
<ul>
	<li>I'm not very familiar about the project/components terms used in Jenkins, however I believe the issue is not about the "parameterized-trigger-plugin" component, but rather about the "build step" used to launch other jobs from within another job.</li>
	<li>The issue is not related to having only 1 executor, but rather to having no available executors when launching too many master/orchestrators jobs that run sub-jobs, causing a deadlock.</li>
</ul>


<p>I propose updating the Jira ticket to reflect the issue as: "Deadlock/Lockup when using "launching builds on other jobs" and "Block until the triggered jobs finish their builds" when no available executors"</p>

<p>From what I understand, this issue occurs in the following scenario:</p>
<ul>
	<li>Node/slave with only 2 (or 1) executors available</li>
	<li>1 upstream job finishes successfully</li>
	<li>2 downstream (orchestrator) jobs are automatically triggered immediatelly by the upstream job when it finishes successfully</li>
	<li>Each downstream job launches other sub-jobs (dependent jobs) that each require their own executor</li>
</ul>


<p>Since the last bullet point requires extra executors, and there are none available, the master/orchestrator jobs will enter a deadlock.</p>

<p>There are several ways to resolve this issue:</p>
<ol>
	<li>(my preference): The launched dependent jobs do not occupy extra executors if run on the same node/slave.</li>
	<li>Distribute the execution of the downstream jobs so that they do not collide and there are available executors for their sub-jobs.</li>
</ol>


<p>A high-level implementation of the above could be:</p>
<ol>
	<li>For the first option, this could be automated by the Jenkins orchestrator. Alternatively, a new flag in the build step could also help.</li>
	<li>For the second way, this could be achieved by adding a flag to the trigger-upstream configuration in the downstream job with a parameter similar to the cron H flag. In other words, instead of immediately executing the downstream jobs, they could be distributed over a set period of time (e.g. within an hour).</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/26959">JENKINS-26959</a>]</i>
